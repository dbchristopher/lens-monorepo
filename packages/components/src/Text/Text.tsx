import { TextDecorationProperty } from 'csstype';
import omit from 'lodash/omit';
import React, { FunctionComponent, Ref } from 'react';
import styled, { css, StyledComponent } from 'styled-components';
import {
  getTextTransform,
  TextTransforms,
  textVariant,
  TextVariants,
} from 'looker-lens-design-tokens';
import { ThemedProps } from '../types';
import { BoxPropsWithout } from '../Box/Box';
import Box from '../Box';

export type TextAlignments = 'left' | 'center' | 'right';
export type TextElements = 'span' | 'p' | 'code';

export interface TextProps
  extends BoxPropsWithout<HTMLSpanElement, 'wrap' | 'is'> {
  /** Base html text element
   *  @default "span"
   */
  is?: TextElements;
  /** Align text */
  align?: TextAlignments;
  /** Set text decoration property */
  decoration?: TextDecorationProperty;
  /** Size mapping from type ramp */
  textTransform?: TextTransforms;
  /** Adjust style of text with more meaning by using an intent */
  variant?: TextVariants;
  /** Should browser insert line breaks within words to prevent text from overflowing its content box  */
  wrap?: boolean;
}

type ComponentType = FunctionComponent<TextProps>;
type StyledComponentType = StyledComponent<ComponentType, TextProps>;

const InternalText: ComponentType = ({
  is = 'span',
  align,
  lineHeight,
  fontSize = 'medium',
  fontWeight,
  ...props
}) => {
  return (
    <Box
      is={is}
      fontSize={fontSize}
      fontWeight={fontWeight}
      lineHeight={lineHeight || fontSize}
      textAlign={align}
      {...omit(props, ['decoration', 'textTransform', 'variant', 'wrap'])}
    >
      {props.children}
    </Box>
  );
};

const TextFactory = React.forwardRef<StyledComponentType, TextProps>(
  (props: TextProps, ref: Ref<StyledComponentType>) => (
    <InternalText ref={ref} {...props} />
  )
);

const Text = styled<ComponentType>(TextFactory)`
  text-decoration: ${(props: ThemedProps<TextProps>) => props.decoration};
  ${(props: ThemedProps<TextProps>) => getTextTransform(props.textTransform)};
  ${(props: ThemedProps<TextProps>) => getWrap(props.wrap || false)};
  ${(props: ThemedProps<TextProps>) => textVariant(props.theme, props.variant)};
`;

const getWrap = (doWrap: boolean) =>
  doWrap
    ? css`
        overflow-wrap: break-word;
      `
    : false;

export default Text;
