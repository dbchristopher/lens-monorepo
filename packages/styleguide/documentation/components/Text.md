```js noeditor
const StatusAndResources = require('../../../styleguide_components/StatusAndResources')
  .StatusAndResources
;<StatusAndResources
  status="experimental"
  figmaURL="https://www.figma.com/file/SmIM3VUQUgo8xb4o7UUxflkT/Typography"
  githubURL="https://github.com/looker/lens/blob/master/src/components/Text/Text.tsx"
  feedbackTitle="Text Component Feedback"
/>
```

### Text Element

By default the `<Text />` component will render text wrapped in a `<span>`, if you need another html element for semantic purposes you can use the `element` property

```js
import { Code } from './Code'
import { Paragraph } from './Paragraph'
import { Text } from './Text'
;<>
  <Text>By default I am a span</Text>
  <Paragraph>Wrapped in a paragraph tag</Paragraph>
  <Code>Wrapped in a code tag</Code>
</>
```

<div class="doc-section-divider"></div>

### Text Size

If you need to adjust the font-size of the rendered text, you can use the `size` property and a value from the [type scale](/#!/Typography)

```js
import { Text } from './Text'
;<>
  <Text fontSize="xxxxlarge">Sized to xxxxlarge</Text>
  <Text fontSize="xxlarge">Sized to xxlarge</Text>
  <Text fontSize="small">Sized to small</Text>
</>
```

<div class="doc-section-divider"></div>

### Weight & Transform

Common patterns for text is to adjust the font weight and transform the text. Below is an example of using the `weight` and `textTransform` properties to modify the rendered text

```js
import { Text } from './Text'
;<>
  <Text fontSize="small" textTransform="upper" fontWeight="semiBold">
    A great story
  </Text>
  <Text fontSize="xxxxlarge" fontWeight="light">
    This is a great headline
  </Text>
  <Text fontSize="small" textTransform="caps">
    Some metadata about this story
  </Text>
</>
```

<div class="doc-section-divider"></div>

### Text Variants

Lens provides a few variants for the `<Text />` component based on common patterns we use in our applications, the `variant` property will change that will change the color the rendered text.

```js
import { Text } from './Text'
;<>
  <Text>Regular Text</Text>
  <Text variant="secondary">Secondary Text</Text>
  <Text variant="subdued">Subdued Text</Text>
  <Text variant="critical">Critical Text</Text>
  <Text variant="positive">Positive Text</Text>
</>
```

<div class="doc-section-divider"></div>

### Text Alignment

You can use the `align` property to change the alignment of the rendered text, below is an example with aligned text put into `<Card />`s.

```js
import { Card, CardContent } from '../Card'
import { Paragraph } from './Paragraph'
;<>
  <Card>
    <CardContent>
      <Paragraph> I am aligned left by default</Paragraph>
    </CardContent>
  </Card>
  <Card>
    <CardContent>
      <Paragraph align="center">
        This is how you can center align Paragraph text
      </Paragraph>
    </CardContent>
  </Card>
  <Card>
    <CardContent>
      <Paragraph align="right">
        This is how you can right align Paragraph text
      </Paragraph>
    </CardContent>
  </Card>
</>
```

### Truncate

At times you may want your `<Paragraph />` or `<Heading />` to truncate instead of the text wrapping, the `truncate` property will do that for you. For a multiline truncate, use `truncateLines` (this is not supported in IE11).

```js
import { Heading } from '../Heading/Heading'
import { Paragraph } from './Paragraph'
;<div style={{ width: '40%' }}>
  <Heading truncate is="h2">
    Hello there I am a heading that will truncate to one line üçïü•ëüç™ü•ì
  </Heading>
  <Paragraph truncateLines={2}>
    And I am some text that is super long and gets truncated at two lines with
    an ellipsis.
  </Paragraph>
</div>
```
