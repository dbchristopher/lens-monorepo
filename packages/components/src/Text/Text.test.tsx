import { Text } from './Text'

import {
  snapshotTestTextComponent,
  snapshotTestTextComponentColor,
  snapshotTestTextComponentDecoration,
  snapshotTestTextComponentFontRamp,
  snapshotTestTextComponentFontWeight,
  snapshotTestTextComponentTransform,
  snapshotTestTextComponentVariant,
  snapshotTestTextComponentWrap,
} from '../../../test/text'

test('A default Text component', () => {
  snapshotTestTextComponent(Text)
})

test('A Text component resized', () => {
  snapshotTestTextComponentFontRamp(Text)
})

test('A Text component weight', () => {
  snapshotTestTextComponentFontWeight(Text)
})

test('A Text component with variant', () => {
  snapshotTestTextComponentVariant(Text)
})

test('A Text component with color', () => {
  snapshotTestTextComponentColor(Text)
})

test('A Text component text transformed', () => {
  snapshotTestTextComponentTransform(Text)
})

test('A Text component wrapped', () => {
  snapshotTestTextComponentWrap(Text)
})

test('A Text component decorated', () => {
  snapshotTestTextComponentDecoration(Text)
})
