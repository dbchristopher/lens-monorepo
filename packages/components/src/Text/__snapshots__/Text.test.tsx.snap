// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A Text component decorated 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  font-size: 1rem;
  line-height: 1.5rem;
}

.c0 {
  -webkit-text-decoration: line-through;
  text-decoration: line-through;
  text-transform: none;
}

<span
  className="c0 c1"
>
  Hello
</span>
`;

exports[`A Text component resized 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  font-size: 2.875rem;
  line-height: 4rem;
}

.c0 {
  text-transform: none;
}

<span
  className="c0 c1"
>
  Hello
</span>
`;

exports[`A Text component text transformed 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  font-size: 1rem;
  line-height: 1.5rem;
}

.c0 {
  text-transform: uppercase;
}

<span
  className="c0 c1"
>
  Hello
</span>
`;

exports[`A Text component weight 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  font-size: 1rem;
  font-weight: 700;
  line-height: 1.5rem;
}

.c0 {
  text-transform: none;
}

<span
  className="c0 c1"
>
  Hello
</span>
`;

exports[`A Text component with color 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  color: #49a9f2;
  font-size: 1rem;
  line-height: 1.5rem;
}

.c0 {
  text-transform: none;
}

<span
  className="c0 c1"
>
  Hello
</span>
`;

exports[`A Text component with variant 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  font-size: 1rem;
  line-height: 1.5rem;
}

.c0 {
  text-transform: none;
  color: #CC1F36;
}

<span
  className="c0 c1"
>
  Hello
</span>
`;

exports[`A Text component wrapped 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  font-size: 1rem;
  line-height: 1.5rem;
}

.c0 {
  text-transform: none;
  overflow-wrap: break-word;
}

<span
  className="c0 c1"
>
  Hello
</span>
`;

exports[`A default Text component 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  font-size: 1rem;
  line-height: 1.5rem;
}

.c0 {
  text-transform: none;
}

<span
  className="c0 c1"
>
  Hello
</span>
`;
